crates/ruff_python_ast/src/nodes.rs

```diff
@@ -1019,19 +1169,191 @@ impl Ranged for ExprStringLiteral {
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct StringLiteralValue {
+    inner: StringLiteralValueInner,
+}
+
+impl StringLiteralValue {
+    /// Creates a new single string literal with the given value.
+    pub fn single(string: StringLiteral) -> Self {
+        Self {
+            inner: StringLiteralValueInner::Single(string),
+        }
+    }
+
+    /// Creates a new string literal with the given values that represents an
+    /// implicitly concatenated strings.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `strings` is less than 2. Use [`StringLiteralValue::single`]
+    /// instead.
+    pub fn concatenated(strings: Vec<StringLiteral>) -> Self {
+        assert!(strings.len() > 1);
+        Self {
+            inner: StringLiteralValueInner::Concatenated(ConcatenatedStringLiteral {
+                strings,
+                value: OnceCell::new(),
+            }),
+        }
+    }
+
+    /// Returns `true` if the string literal is implicitly concatenated.
+    pub const fn is_implicit_concatenated(&self) -> bool {
+        matches!(self.inner, StringLiteralValueInner::Concatenated(_))
+    }
+
+    /// Returns `true` if the string literal is a unicode string.
+    ///
+    /// For an implicitly concatenated string, it returns `true` only if the first
+    /// string literal is a unicode string.
+    pub fn is_unicode(&self) -> bool {
+        self.parts().next().map_or(false, |part| part.unicode)
+    }
+
+    /// Returns an iterator over all the [`StringLiteral`] parts contained in this value.
+    pub fn parts(&self) -> impl Iterator<Item = &StringLiteral> {
+        match &self.inner {
+            StringLiteralValueInner::Single(value) => Left(std::iter::once(value)),
+            StringLiteralValueInner::Concatenated(value) => Right(value.strings.iter()),
+        }
+    }
+
+    /// Returns an iterator over all the [`StringLiteral`] parts contained in this value
+    /// that allows modification.
+    pub(crate) fn parts_mut(&mut self) -> impl Iterator<Item = &mut StringLiteral> {
+        match &mut self.inner {
+            StringLiteralValueInner::Single(value) => Left(std::iter::once(value)),
+            StringLiteralValueInner::Concatenated(value) => Right(value.strings.iter_mut()),
+        }
+    }
+
+    /// Returns the concatenated string value as a [`str`].
+    pub fn as_str(&self) -> &str {
+        match &self.inner {
+            StringLiteralValueInner::Single(value) => value.as_str(),
+            StringLiteralValueInner::Concatenated(value) => value.as_str(),
+        }
+    }
+}
+
+impl PartialEq<str> for StringLiteralValue {
+    fn eq(&self, other: &str) -> bool {
+        self.as_str() == other
+    }
+}
+
+impl PartialEq<String> for StringLiteralValue {
+    fn eq(&self, other: &String) -> bool {
+        self.as_str() == other
+    }
+}
+
+impl Deref for StringLiteralValue {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        self.as_str()
```

Comment:
I don't know that I'm a huge fan of `deref` here since this can actually be expensive and require an allocation. How much trouble would it be to remove this and require explicit `as_str()`?

Reply:
It's not too much as I just need to go through the references and add the method call which should be simple with `rust-analyzer`.

I get your point and I think having an explicit call is worth the small effort.

Reply:
#8826